src/mame/misc/babysuprem.cpp

```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
+	PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_BUTTON7)
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_BUTTON8)
+
 INPUT_PORTS_END
 
 void bsuprem_state::bsuprem(machine_config &config)
 {
 	// Basic machine hardware
 	Z80(config, m_maincpu, 2.4576_MHz_XTAL); // divider not verified
+	m_maincpu->set_addrmap(AS_PROGRAM, &bsuprem_state::bsuprem_map);
+	m_maincpu->set_periodic_int(FUNC(bsuprem_state::mains_irq), attotime::from_hz(100));
+
+	NVRAM(config, "nvram", nvram_device::DEFAULT_ALL_0);
+
+	HOPPER(config, m_hopper_5, attotime::from_msec(50), TICKET_MOTOR_ACTIVE_HIGH, TICKET_STATUS_ACTIVE_HIGH);
+	HOPPER(config, m_hopper_25, attotime::from_msec(50), TICKET_MOTOR_ACTIVE_HIGH, TICKET_STATUS_ACTIVE_HIGH);
+	HOPPER(config, m_hopper_100, attotime::from_msec(50), TICKET_MOTOR_ACTIVE_HIGH, TICKET_STATUS_ACTIVE_HIGH);
 
 	// Sound hardware
 	SPEAKER(config, "mono").front_center();
 	AY8910(config, m_psg[0], 2.4576_MHz_XTAL); // divider not verified
 	m_psg[0]->add_route(ALL_OUTPUTS, "mono", 0.30);
 	AY8910(config, m_psg[1], 2.4576_MHz_XTAL); // divider not verified
 	m_psg[1]->add_route(ALL_OUTPUTS, "mono", 0.30);
+
+	m_psg[0]->port_a_write_callback().set(FUNC(bsuprem_state::ay8910_u1_porta_w));
+	m_psg[0]->port_b_write_callback().set(FUNC(bsuprem_state::ay8910_u1_portb_w));
+	m_psg[1]->port_a_write_callback().set(FUNC(bsuprem_state::ay8910_u2_porta_w));
+	m_psg[1]->port_b_write_callback().set(FUNC(bsuprem_state::ay8910_u2_portb_w));
+}
+
+void bsuprem_state::bsuprem_map(address_map &map)
+{
+	map(0x0000, 0x0fff).rom();
+	map(0x1000, 0x1fff).rw(FUNC(bsuprem_state::ay8910_w_dsw1_switches2_r), FUNC(bsuprem_state::ay8910_w));
+	map(0x2000, 0x2fff).portr("SWITCHES");
+	map(0x3000, 0x3fff).r(FUNC(bsuprem_state::dsw2_u13_switches_r));
+	map(0x4000, 0x4fff).rom().region("maincpu", 0x2000);
+	map(0x5000, 0x5fff).w(FUNC(bsuprem_state::u9_w));
+	map(0x6000, 0x6fff).w(FUNC(bsuprem_state::u10_w));
+	map(0x7000, 0x77ff).mirror(0x0800).rw(FUNC(bsuprem_state::nvram_r), FUNC(bsuprem_state::nvram_w));
+
+	map(0x8000, 0x8fff).rom().region("maincpu", 0x1000);
+	map(0xc000, 0xcfff).rom().region("maincpu", 0x3000);
+}
+
+void bsuprem_state::machine_reset()
+{
+	m_irq_state = 0;
+	m_u1_porta_data = 0;
+	m_u10_data = 0;
+}
+
+void bsuprem_state::machine_start()
+{
+	m_digits.resolve();
+	m_lamps.resolve();
+	m_buttons.resolve();
+	m_nixie.resolve();
+
+	m_nvram8 = std::make_unique<uint8_t[]>(2048);
+	m_nvram->set_base(m_nvram8.get(),2048);
+
+	save_item(NAME(m_irq_state));
+	save_item(NAME(m_u1_porta_data));
+	save_item(NAME(m_u10_data));
+}
+
+void bsuprem_state::nvram_w(offs_t offset, uint8_t data)
+{
+	// according to the schematic the ram c/s is disabled unless
+	// bit 6 of u10 is 1. However, implementing this makes it impossible
+	// to enter test/setup. As there are a number of other errors on
+	// the schematic it's likely this is wrong.
+	m_nvram8[offset] = data;
+}
+
+uint8_t bsuprem_state::nvram_r(offs_t offset)
+{
+	return m_nvram8[offset];
+}
```

Comment:
Do you actually need this trampoline?  Could you install it as ram with `.share(...)` so you can get a convenient pointer to it, and then not need to allocate/save it yourself either?